File: frontend/src/api/inventory.ts
import client from './client';

export interface InventoryItem {
  canonical_item_id: string;
  canonical_name: string;
  category: string;
  current_quantity: number;
  unit: string;
  threshold_quantity: number;
  daily_consumption_estimate: number;
  estimated_days_left: number | null;
  status: 'ok' | 'low' | 'critical' | 'out_of_stock';
  last_updated_at: string;
  last_receipt_at?: string;
  reminder_enabled: boolean;
  last_reminder_sent_at?: string;
}

export interface InventoryDetail {
  item: InventoryItem;
  recent_events: Array<{
    id: string;
    event_type: string;
    delta_quantity: number;
    unit: string;
    created_at: string;
    source: { type: string; referenceId?: string };
  }>;
}

export const inventoryApi = {
  list: (params?: { status?: string; category?: string; sort_by?: string }) =>
    client.get<never, InventoryItem[]>('/api/v1/inventory', { params }),

  lowStock: () =>
    client.get<never, InventoryItem[]>('/api/v1/inventory/low-stock'),

  getItem: (itemId: string) =>
    client.get<never, InventoryDetail>(`/api/v1/inventory/${itemId}`),

  adjust: (itemId: string, data: { current_quantity: number; unit: string; reason?: string }) =>
    client.patch<never, InventoryItem>(`/api/v1/inventory/${itemId}`, data),

  reset: () =>
    client.delete<never, { states_deleted: number; events_deleted: number }>('/api/v1/inventory/reset'),
};

File: app/models/inventory.py
"""Inventory domain models."""

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


InventoryStatus = Literal["ok", "low", "critical", "out_of_stock"]
EventType = Literal[
    "purchase",
    "daily_decrement",
    "manual_adjustment",
    "receipt_deleted",
    "consumption_correction",
    "waste_discarded",
]


class InventoryMetadata(BaseModel):
    """Extra metadata on an inventory state document."""
    source_confidence: Optional[float] = None
    last_purchase_store: Optional[str] = None
    last_purchase_price: Optional[float] = None


class InventoryState(BaseModel):
    """Current pantry state for one user-item pair.

    Matches the MongoDB `inventory_state` collection.
    """
    id: Optional[str] = Field(None, alias="_id")
    user_id: str
    canonical_item_id: str
    canonical_name: str
    category: Optional[str] = None
    current_quantity: float = 0.0
    unit: str = "count"
    threshold_quantity: float = 0.0
    daily_consumption_estimate: float = 0.0
    estimated_days_left: Optional[float] = None
    status: InventoryStatus = "ok"
    last_updated_at: datetime = Field(default_factory=datetime.utcnow)
    last_receipt_at: Optional[datetime] = None
    reminder_enabled: bool = True
    last_reminder_sent_at: Optional[datetime] = None
    metadata: InventoryMetadata = Field(default_factory=InventoryMetadata)

    model_config = {"populate_by_name": True}


class EventSource(BaseModel):
    """Source reference for an inventory event."""
    type: str = ""
    reference_id: Optional[str] = None


class InventoryEvent(BaseModel):
    """Append-only event log entry for inventory changes.

    Matches the MongoDB `inventory_events` collection.
    """
    id: Optional[str] = Field(None, alias="_id")
    user_id: str
    canonical_item_id: str
    canonical_name: str
    event_type: EventType
    delta_quantity: float
    unit: str
    created_at: datetime = Field(default_factory=datetime.utcnow)
    source: EventSource = Field(default_factory=EventSource)
    meta: dict[str, Any] = Field(default_factory=dict)

    model_config = {"populate_by_name": True}
